docs/reference/modules/discovery/fault-detection.asciidoc

```diff
@@ -31,3 +31,297 @@ The elected master node will also remove nodes from the cluster if nodes are una
+the check request message to the follower.
+
+. The master's `TransportService` running on thread
+`elasticsearch[master][transport_worker][T#2]` passes the check request message
+onto the operating system.
+
+. The operating system on the master converts the message into one or more
+packets and sends them out over the wire.
+
+. Miscellaneous routers/firewalls/other devices between master & follower node
+forward the packets, possibly fragmenting or defragmenting them on the way.
+
+. The operating system on the follower node receives the packets and notifies
+{es} that they've been received.
+
+. The follower's `TransportService` running on thread
+`elasticsearch[follower][transport_worker][T#3]` reads the incoming packets and
+reconstructs and processes the check request. Usually the check succeeds
+straight away and the same thread immediately constructs a response and passes
+it back to the operating system.
+
+. If, rarely, the check didn't succeed straight away (e.g. an election started
+recently) then:
+
+.. The follower's `FollowerChecker` running on thread
+`elasticsearch[follower][cluster_coordination][T#4]` processes the request,
+constructs a response, and tells the `TransportService` to send the response
+back to the master.
+
+.. The follower's `TransportService` running on thread
+`elasticsearch[follower][transport_worker][T#3]` passes the response on to the
+operating system.
+
+. The operating system on the follower converts the response into one or more
+packets and sends them out over the wire.
+
+. Miscellaneous routers/firewalls/other devices between master & follower
+forward the packets, possibly fragmenting or defragmenting them on the way.
+
+. The operating system on the master receives the packets and notifies {es}
+that they've been received.
+
+. The master's `TransportService` running on thread
+`elasticsearch[master][transport_worker][T#2]` reads the incoming packets,
+reconstructs the check response and processes it as long as the check didn't
+already time out.
+
+This means that there's a lot of different things that can delay the completion
+of a check for long enough for it to time out. Here are some examples for each
+step:
+
+. There may be a long garbage collection (GC) or virtual machine (VM) pause
+after passing the check request to the `TransportService`.
+
+. There may be a long wait for the specific `transport_worker` thread to become
+available, or there may be a long GC or VM pause before passing the check
+request onto the operating system.
+
+. A system fault (e.g. broken network card) on the master may delay sending the
+message over the wire, possibly indefinitely.
+
+. Intermediate devices may delay, drop, or corrupt packets along the way. The
+OS on the master will wait and retransmit any unacknowledged or corrupted
+packets up to `net.ipv4.tcp_retries2` times; we recommend
+<<system-config-tcpretries,reducing this value>> since the default represents a
+very long delay.
```

Comment:
```suggestion
. Intermediate devices may delay, drop, or corrupt packets along the way. The
operating system for the master will wait and retransmit any unacknowledged or
corrupted packets up to `net.ipv4.tcp_retries2` times. We recommend
<<system-config-tcpretries,reducing this value>> since the default represents a
very long delay.
```